sdk/src/client/node_api/indexer/query_parameters.rs

```diff
@@ -8,69 +8,235 @@ use std::fmt;
 
+pub trait QueryParameterHelper {
+    /// Converts parameters to a single String.
+    fn to_query_string(&self) -> Option<String>;
+    fn replace_cursor(&mut self, cursor: String);
+}
+
+macro_rules! impl_query_parameters_methods {
+    ($name:ty, $inner_type:ty) => {
+        impl $name {
+            /// Creates a hashset from a provided vec of query parameters.
+            #[must_use]
+            pub fn new(query_parameters: impl Into<Vec<$inner_type>>) -> Self {
+                let mut query_parameters = query_parameters.into();
+                query_parameters.sort_unstable_by_key(<$inner_type>::kind);
+                query_parameters.dedup_by_key(|qp| qp.kind());
+
+                Self(query_parameters)
+            }
+            /// Replaces or inserts an enum variant in the QueryParameters.
+            pub fn replace(&mut self, query_parameter: $inner_type) {
+                match self
+                    .0
+                    .binary_search_by_key(&query_parameter.kind(), <$inner_type>::kind)
+                {
+                    Ok(pos) => self.0[pos] = query_parameter,
+                    Err(pos) => self.0.insert(pos, query_parameter),
+                }
+            }
+            /// Creates new empty QueryParameters.
+            pub fn empty() -> Self {
+                Self(Vec::new())
+            }
+        }
+        impl QueryParameterHelper for $name {
+            /// Converts parameters to a single String.
+            fn to_query_string(&self) -> Option<String> {
+                query_tuples_to_query_string(self.0.iter().map(|q| Some(q.to_query_tuple())))
+            }
+            fn replace_cursor(&mut self, cursor: String) {
+                self.replace(<$inner_type>::Cursor(cursor));
+            }
+        }
+
+        impl fmt::Display for $inner_type {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                let query_tuple = self.to_query_tuple();
+                write!(f, "{}={}", query_tuple.0, query_tuple.1)
+            }
+        }
+    };
+}
+
 /// Query parameters for output_id requests.
 #[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct QueryParameters(Vec<QueryParameter>);
+pub struct OutputsQueryParameters(Vec<OutputsQueryParameter>);
 
-impl QueryParameters {
-    /// Creates a hashset from a provided vec of query parameters.
-    #[must_use]
-    pub fn new(query_parameters: impl Into<Vec<QueryParameter>>) -> Self {
-        let mut query_parameters = query_parameters.into();
-        query_parameters.sort_unstable_by_key(QueryParameter::kind);
-        query_parameters.dedup_by_key(|qp| qp.kind());
+impl_query_parameters_methods!(OutputsQueryParameters, OutputsQueryParameter);
 
-        Self(query_parameters)
-    }
+/// Query parameter for output requests.
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+#[non_exhaustive]
+pub enum OutputsQueryParameter {
```

Comment:
Should these just be structs?

Reply:
Because each of the variants can be there at the same time and only once? Or why do you think would a struct be better?

Reply:
Yeah exactly. Feels like a struct to me

Reply:
Changed to a struct :+1: